File: src/main/java/Test01.java
import java.util.TreeMap;

public class Test01 {
    public static void main(String[] args) {
        User user1 = new User("jack", 11);
        User user2 = new User("tom", 11);
        TreeMap<User, String> map = new TreeMap<>();
        map.put(user1,"a");
        map.put(user2,"b");
        System.out.println(map);
    }
}

File: src/main/scala/Test/Test01.scala
package Test

object Test01 {
  def main(args: Array[String]): Unit = {
    def userCompare[T : Ordering](a : T,b : T): T ={
      val ord: Ordering[T] = implicitly[Ordering[T]]
      ord.max(a,b)
    }
    implicit val ord : Ordering[User] = new Ordering[User] {
      override def compare(x: User, y: User): Int = x.age - y.age
    }

    println(userCompare(User("jack", 10), User("tom", 12)))
  }
}
case class User(name : String ,age : Int ){

}
